sw/device/lib/crypto/impl/ecc.c

```diff
@@ -644,28 +663,277 @@ crypto_status_t otcrypto_ecdsa_verify_async_finalize(
+      OTCRYPTO_TRY_INTERPRET(
+          internal_ecdh_p256_keygen_finalize(private_key, public_key));
+      return kCryptoStatusOK;
+    case kEccCurveTypeNistP384:
+      OT_FALLTHROUGH_INTENDED;
+    case kEccCurveTypeBrainpoolP256R1:
+      OT_FALLTHROUGH_INTENDED;
+    case kEccCurveTypeCustom:
+      // TODO: Implement support for other curves.
+      return kCryptoStatusNotImplemented;
+    default:
+      return kCryptoStatusBadArgs;
+  }
+
+  // Should never get here.
+  HARDENED_UNREACHABLE();
+  return kCryptoStatusFatalError;
+}
+
+/**
+ * Start an ECDH shared key generation operation for curve P-256.
+ *
+ * @param private_key Private key for key exchange.
+ * @param public_key Public key for key exchange.
+ * @return OK or error.
+ */
+static status_t internal_ecdh_p256_start(
+    const crypto_blinded_key_t *private_key,
+    const ecc_public_key_t *public_key) {
+  if (private_key->config.hw_backed != kHardenedBoolFalse) {
+    // TODO: Implement support for sideloaded keys.
+    return OTCRYPTO_NOT_IMPLEMENTED;
+  }
+
+  // Check the private key size.
+  HARDENED_TRY(p256_private_key_length_check(private_key));
+
+  // Get pointers to the individual shares within the blinded key.
+  uint32_t *share0;
+  uint32_t *share1;
+  HARDENED_TRY(keyblob_to_shares(private_key, &share0, &share1));
+
+  // Copy the shares into a P256-specific struct.
+  p256_masked_scalar_t sk;
+  memcpy(sk.share0, share0, sizeof(sk.share0));
+  memcpy(sk.share1, share1, sizeof(sk.share1));
+
+  // Check the public key size.
+  HARDENED_TRY(p256_public_key_length_check(public_key));
+
+  // Copy the public key into a P256-specific struct.
+  p256_point_t pk;
+  memcpy(pk.x, public_key->x.key, sizeof(pk.x));
+  memcpy(pk.y, public_key->y.key, sizeof(pk.y));
+
+  // Check the lengths of caller-allocated buffers.
+  HARDENED_TRY(p256_private_key_length_check(private_key));
+  HARDENED_TRY(p256_public_key_length_check(public_key));
+
+  // Note: This operation wipes DMEM after retrieving the key, so if an error
+  // occurs after this point then the keys would be unrecoverable. This should
+  // be the last potentially error-causing line before returning to the caller.
+  return ecdh_p256_shared_key_start(&sk, &pk);
 }
 
 crypto_status_t otcrypto_ecdh_async_start(
     const crypto_blinded_key_t *private_key, const ecc_public_key_t *public_key,
     const ecc_curve_t *elliptic_curve) {
-  // TODO: Connect ECDH operations to API.
-  return kCryptoStatusNotImplemented;
+  if (private_key == NULL || public_key == NULL || elliptic_curve == NULL) {
+    return kCryptoStatusBadArgs;
+  }
+
+  // Check the integrity of the private key.
+  if (integrity_blinded_key_check(private_key) != kHardenedBoolTrue) {
+    return kCryptoStatusBadArgs;
+  }
+
+  // Check the private key configuration.
+  OTCRYPTO_TRY_INTERPRET(
+      key_config_check(elliptic_curve, &private_key->config, kKeyModeEcdh));
+
+  // Select the correct signing operation and start it.
```

Comment:
nit: `// Select the correct ecdh operation and start it`? Same for other applicable comments.